### ISSUE TITLE
profiling.h conflicts with Qt

### ISSUE THREAD

**Issue by zappaz00** [oneapi-src/oneTBB#547]
profiling.h has method with name "emit" and strongly conficts with Qt keyword emit() used for signal/slot mechanism; Please, rename this method (problem starts from one api and absents in parallel studio). Qt won't fix it. Now developer can't correctly use Qt and TBB together.
![image](https://user-images.githubusercontent.com/18219766/130943646-749df6e3-306f-4c43-aced-a2cece5c9a8d.png)

**Comment by vlserov** [comment]
profiling.h is a public API and changes might affect other users.
Using TBB with QT_NO_KEYWORDS is the right way. But if  project is huge and it's complicated, consider to use workaround suggested for Boost Signals: [boost signals doc](https://www.boost.org/doc/libs/1_53_0/doc/html/signals/s04.html)
```
#ifndef Q_MOC_RUN
#if defined(emit)
    #undef emit
    #include <tbb/tbb.h>
    #define emit // restore the macro definition of "emit", as it was defined in gtmetamacros.h
#else
    #include <tbb/tbb.h>
#endif // defined(emit)
#endif // Q_MOC_RUN
```

**Comment by alexey-katranov** [comment]
> rename this method (problem starts from one api and absents in parallel studio)

[The same code was located](https://github.com/oneapi-src/oneTBB/blob/tbb_2020/include/tbb/tbb_profiling.h#L340) in tbb_profiling.h with TBB that was a part of Intel Parallel Studio. Maybe oneTBB includes profiling.h in some additional place. What headers do you include from oneTBB?

@vlserov  Perhaps, we can try to think about some workarounds or options to hide this code when user requests.

**Comment by keithel** [comment]
@zappaz00 `QT_NO_KEYWORDS` does not need to be defined when compiling Qt (which one comment in #309 mentioned). You just need to define `QT_NO_KEYWORDS` before including any Qt headers, and then replace you use of `signals`, `signal`, `slots`, `slot, and `emit` with `Q_SIGNALS`, `Q_SIGNAL`, `Q_SLOTS`, `Q_SLOT`, and `Q_EMIT`, respectively.

If you're not actually using the `emit` function in OneTBB, in a file that includes some OneTBB thing that defines `emit`, then I think you can just get away with including the OneTBB headers before any Qt ones, and everything will compile ok. If you're using the OneTBB `emit` function, however, then you need to go with the `QT_NO_KEYWORDS` approach.

You can apply `QT_NO_KEYWORDS` across your whole app, or just apply it targeted, to each affected source file.

See https://doc.qt.io/qt-6/signalsandslots.html#using-qt-with-3rd-party-signals-and-slots for more information.

**Comment by mnikelsky** [comment]
You don´t even have to rename it, just adding something like
#pragma push_macro("emit")
#undef emit
....
#pragma pop_macro("emit")

to profiling.h fixes the issue. Just add a TBB define we can set to OneTBB so we can decide whether we want to have this or not. Having to patch the tbb each time a new version comes out is annoying.

**Comment by gtkramer** [comment]
TBB and Qt ultimately need to be able to co-exist in an application.  Using `-DQT_NO_KEYWORDS` after doing a find and replace operation on a code base of any size is a great start, but it will only get you so far.  It will work with Qt and VTK, but not with other extensions like qicstable, which need to be fixed.

**Comment by kikaxa** [comment]
A bit of shame that so many implementation details leak and break other libraries(concerns both qt and less tbb).
It is not possible to work around in other huge library(ies) that pull in tbb somewhere, it is not reasonable to change standard qt code that is using emit.

best fix is to hide as much implementation from headers. 
in our case, removing the body of event in a patch is enough to make it work.
